emds: adapted documentation (#23518)
nrfconnect/sdk-nrf: doc/nrf/libraries/others/emds.rst

**************
The :kconfig:option:`CONFIG_EMDS` Kconfig option enables the emergency data storage.

The EMDS is implemented to use the persistent memory in form of storage partitions, which can be either flash or RRAM.
The static partition manager file :file:`pm.yml.emds` is required for defining the storage partitions.

> The storage partitions should be aligned to write block size, which is 4 bytes for flash and 16 bytes for RRAM.

Reviewer: ```suggestion
The storage partitions must be aligned to the write block size, which is 4 bytes for flash and 16 bytes for RRAM.
```

The EMDS mandates at least two storage partitions to allow the fallback mechanism.

> The application must initialize storage partitions by using the :c:func:`emds_init` function.

Author: I do not think we need to specify that these are EMDS partitions. This is EMDS documentation. We need to specify belongingness of partitions only if they belong to another subsystem.

> The partition storage space must be larger than the combined data stored by the application.

Reviewer 2: Regarding twice: The issue that I see is, it uses the phrase "storage space must be larger than the combined data stored by the application". It does not tell the user that partition sizing should also include snapshot metadata size. 

Alternatively:
```suggestion
The partition storage space for each partition must be larger than the combined data stored by the application, you can find out how much storage space is needed by calling :c:func:`emds_store_size_get` API.
```

Allocating a larger storage area will demand more resources, but also increase the life expectancy of the persistent memory.
The chosen size should reflect the amount of data stored, the available persistent memory and how the application calls the :c:func:`emds_store` function.

The memory location that is going to be stored must be added on initialization.
All memory areas must be provided through entries containing an ID, data pointer, and data length, using the :c:func:`emds_entry_add` function and the :c:macro:`EMDS_STATIC_ENTRY_DEFINE` macro.
Entries to be stored when the emergency data storage is triggered need their own unique IDs that are not changed after a reboot.
The entries are stored in the partition in form of snapshots.
The snapshot is a packed copy of all entries that are registered in the EMDS stored in the data area of the partition.
Additionally, each snapshot is described by the snapshot metadata that is stored in the metadata area of the partition.

> When all entries are added, the :c:func:`emds_load` function starts a partition scanning procedure to find the latest snapshot.

Author: removed `phase`. We do not have phases.

The EMDS scans the metadata area of both partitions.

> The :kconfig:option:`CONFIG_EMDS_SCANNING_FAILURES` option configures the depth of the scanning procedure.

Reviewer: ```suggestion
The :kconfig:option:`CONFIG_EMDS_SCANNING_FAILURES` Kconfig option configures the depth of the scanning procedure.
```

> If the scanning procedure fails the configured number of times in the row, the scanning is stopped, since highly likely there is no stored metadata anymore.

Reviewer: ```suggestion
If the scanning procedure fails the configured number of times in a row, the scanning is stopped, because most likely, there is no stored metadata left.
```

> The scanning procedure gathers the :kconfig:option:`CONFIG_EMDS_MAX_CANDIDATES` most recent snapshots from the partition metadata area.

Reviewer: ```suggestion
The scanning procedure gathers the most recent snapshots (set by :kconfig:option:`CONFIG_EMDS_MAX_CANDIDATES`) from the partition metadata area.
```

> After the scanning procedure is completed, the EMDS checks data integrity on gathered candidates, starting from the latest stored on the partition.

Reviewer: ```suggestion
After the scanning procedure is completed, the EMDS checks data integrity on the gathered candidates, starting from the latest one stored on the partition.
```

> If data integrity checks pass, it restores the entries into the memory areas from the persistent memory.

Reviewer: ```suggestion
If the data integrity checks pass, the entries are restored into the memory areas from the persistent memory.
```

> If candidates are not found or the data integrity does not pass, the function will return an error.

Reviewer: ```suggestion
If candidates are not found or the data integrity check does not pass, the function returns an error.
```

> Note that the :c:func:`emds_load` function will return error code if run it on the empty partitions.

Reviewer: ```suggestion
The :c:func:`emds_load` function returns an error code if it runs on the empty partitions.
```

> After restoring the previous data, the application must run the :c:func:`emds_prepare` function to prepare the storage area for receiving new entries.

Reviewer 2: ```suggestion
After restoring the previous data, the application must call the :c:func:`emds_prepare` function to prepare the storage area for receiving new entries.
```
?

> Initially, the EMDS will try allocation of the storage area for the new entries in the partition that has the most recent snapshot to utilize partition capacity at maximum.

Reviewer: ```suggestion
Initially, the EMDS tries to allocate the storage area for the new entries in the partition that has the most recent snapshot to use the partition capacity at maximum.
```

> Initially, the EMDS tries to allocate the storage area for the new entries in the partition that has the most recent snapshot to use the partition capacity at maximum.

Reviewer 2: minor: optional:

```suggestion
Initially, the EMDS library tries to allocate the storage area for the new entries in the partition that has the most recent snapshot in order to use the full partition capacity.
```
I think similar updates are needed at other places as well. Because sometimes this doc uses "EMDS" to mean storage area, sometimes it uses "EMDS" to mean library, and sometimes to mean a specific mechanism. That causes a bit of confusion when someone is reading this for the first time.

> However, if the partition is full, the EMDS will try to allocate the storage area in the other partition.

Reviewer: ```suggestion
However, if the partition is full, the EMDS tries to allocate the storage area in the other partition.
```

> If the partition is full, the EMDS will erase the whole partition to get space for new snapshot if it is applicable for the persistent memory.

Reviewer: ```suggestion
If the partition is full, the EMDS erases the entire partition to get space for a new snapshot if it is applicable for the persistent memory.
```

> If the partition is full, the EMDS erases the entire partition to get space for a new snapshot if it is applicable for the persistent memory.

Reviewer 2: ```suggestion
If the other partition is full, the EMDS erases the entire partition to get space for a new snapshot if erase is needed, depending on the type of persistent memory. Devices with flash memories require explicit erase, whereas devices with RRAM do not need explicit erase. The EMDS library handles this internally.
```

> The EMDS will never erase the partition that has the most recent snapshot found during the partitions scanning procedure.

Reviewer: ```suggestion
The EMDS never erases the partition that has the most recent snapshot found during the partitions scanning procedure.
```

The storage is done in deterministic time, so it is possible to know how long it takes to store all registered entries.
However, this is chip-dependent, so it is important to measure the time.
Find timing values under the "Electrical specification" section for the non-volatile memory controller in the Product Specification for the relevant SoC or the SiP you are using.
For example, for the nRF52840 SiP, see the `nRF52840 Product Specification`_ page.
The storing is performed by chunks of data with a size of 16 bytes.

> The storing time is determined by the chunk preparation time and the flash writing time and depends on both the number of stored data bytes (both data and metadata) as well as the number of chunks.

Reviewer: ```suggestion
The storing time is determined by the chunk preparation time and the flash writing time, and depends on both the number of stored data bytes (both data and metadata) as well as the number of chunks.
```

The following (non-public) Kconfig options are needed for the time estimation:

* :kconfig:option:`CONFIG_EMDS_FLASH_TIME_WRITE_ONE_WORD_US`
* :kconfig:option:`CONFIG_EMDS_CHUNK_PREPARATION_TIME_US`

> You can tune these options to estimate the writing time more precisely, but they do not change the real time required for storing the snapshot.

Reviewer 2: ```suggestion
You can tune these options to influence the estimation of the writing time (see :c:func:`emds_store_time_get`), but they do not change the actual time needed for storing the snapshot. So, it is recommended to consider worst case values if these options need to be changed.
```

The :kconfig:option:`CONFIG_EMDS_SCANNING_FAILURES` Kconfig option configures the depth of the scanning procedure.
The depth is the number of times the scanning procedure can fail in a row before it stops, because most likely, there is no stored metadata left.
The scanning procedure is performed by reading the metadata area of the partition and checking the integrity of the metadata.
The scanning procedure gathers the most recent snapshots (set by :kconfig:option:`CONFIG_EMDS_MAX_CANDIDATES`) from the partition metadata area.
After the scanning procedure is completed, the EMDS checks data integrity on the gathered candidates, starting from the latest one stored on the partition.
If the data integrity checks pass, the entries are restored into the memory areas from the persistent memory.
If candidates are not found or the data integrity check does not pass, the function returns an error.
The :c:func:`emds_load` function returns an error code if it runs on the empty partitions.

After restoring the previous data, the application must call the :c:func:`emds_prepare` function to prepare the storage area for receiving new entries.

> Initially, the EMDS library tries to allocate the storage area for the new entries in the partition that has the most recent snapshot in order to use the partition capacity at maximum.

Reviewer 2: Minor: 
```suggestion
Initially, the EMDS library tries to allocate the storage area for the new entries in the partition that has the most recent snapshot in order to use the maximum partition capacity.
```
'at maximum' at the end sounds quite informal.

The easiest way of computing an estimate of the time required to store all entries, in a worst case scenario, is to call the :c:func:`emds_store_time_get` function.
This function returns a worst-case storage time estimate in microseconds (µs) for a given application.

> To make this work, you need to set the Kconfig options :kconfig:option:`CONFIG_EMDS_FLASH_TIME_WRITE_ONE_WORD_US` and :kconfig:option:`CONFIG_EMDS_CHUNK_PREPARATION_TIME_US` as described in the `Implementation`_ section.

Reviewer 2: ```suggestion
To make this work, you need to determine and set the Kconfig options :kconfig:option:`CONFIG_EMDS_FLASH_TIME_WRITE_ONE_WORD_US` and :kconfig:option:`CONFIG_EMDS_CHUNK_PREPARATION_TIME_US` as described in the `Implementation`_ section for your platform.
```

Actually, I'm wondering now, if this is a right guidance. Since preparation time is not going to change per platform. For given clock frequency and for given chip architecture, the worst case time will be fixed. It is after all a program execution, and that won't change with temperature. We already specify this in the KConfig for 52, 53, and 54. So, I think asking people to measure this, can be too much? Since this is not easy.

Should we remove `and :kconfig:option:`CONFIG_EMDS_CHUNK_PREPARATION_TIME_US`` from this sentence?

Author: Since we provide sample how to calculate writing time where this option participates, we need to keep it in documentation with option clarification. Otherwise, it will look like magic value.

The :c:func:`emds_store_time_get` function estimates the required worst-case time to store :math:`n` entries using the following formula:

.. math::

>    t_\text{store} = t_\text{word}\lceil\frac{s_\text{meta}}{s_\text{word}}\right\rceil + t_\text{word} \sum_{i = 1}^n (\left\lceil\frac{s_i}{s_\text{word}}\right\rceil \right) + t_\text{chunk} \sum_{i = 1}^n (\left\lceil\frac{s_i}{s_\text{chunk}}\right\rceil \right)

Reviewer 2: Do you know if this is rendered somewhere?
